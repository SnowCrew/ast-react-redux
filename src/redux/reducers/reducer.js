

const initialState = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload,
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const id = action.payload;

      const itemInd = state.itemsInCart.findIndex(item => item.id === id);

      if (itemInd >= 0) {
        const elemInState = state.itemsInCart.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: ++elemInState.counterIdentical
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart.slice(0, itemInd),
            newElem,
            ...state.itemsInCart.slice(itemInd + 1)
          ],
          total: (Number(state.total) + Number(newElem.price)).toFixed(2)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem = {
        title: item.title,
        price: item.price,
        image: item.image,
        id: item.id,
        counterIdentical: 1,
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: (Number(state.total) + Number(newItem.price)).toFixed(2)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload;
      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: state.total - (itemDel.price * itemDel.counterIdentical)
      };

    default:
      return state;
  }
}

export default reducer;