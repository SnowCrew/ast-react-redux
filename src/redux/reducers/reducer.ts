import { IAction } from '../actions/actions';


const initialState:IStore = {
  products: [],
  loading: true,
  error: false,
  itemsInCart: [],
  total: 0,
  totalQuantity: 0,
  auth: false,
  admin: false
};

export interface IProduct {
  id: number,
  title:string,
  price:number,
  category:string,
  description:string,
  image:string
}

export interface IProductInCart extends IProduct {
counterIdentical:number,
}

export interface IStore {
  products: IProduct[],
  loading: boolean,
  error: boolean,
  itemsInCart: IProductInCart[],
  total: number,
  totalQuantity: number,
  auth: boolean,
  admin: boolean
};

const reducer = (state:IStore = initialState, action:IAction):IStore => {
  switch (action.type) {
    case 'MENU_LOADED':
      return {
        ...state,
        products: action.payload as IProduct[],
        loading: false,
        error: false
      };
    case 'MENU_REQUESTED':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: false
      };
    case 'MENU_ERROR':
      return {
        ...state,
        products: state.products,
        loading: true,
        error: true
      };
    case "ITEM_ADD_TO_CART":
      const [id, quantity] = action.payload as [number, number];
      const itemInd = state.itemsInCart?.findIndex(item => item.id === id);

      if (itemInd >= 0 ) {
        const elemInState = state.itemsInCart?.find(item => item.id === id);
        const newElem = {
          ...elemInState,
          counterIdentical: (Number(elemInState?.counterIdentical) + Number(quantity)),
        }
        return {
          ...state,
          itemsInCart: [
            ...state.itemsInCart!.slice(0, itemInd),
            newElem,
            ...state.itemsInCart!.slice(itemInd! + 1)
          ] as IProductInCart[],
          total: Number((Number(state.total) + (Number(newElem.price) * Number(quantity))).toFixed(2)),
          totalQuantity: Number(state.totalQuantity) + Number(quantity)
        }
      }

      const item = state.products.find(item => item.id === id);
      const newItem:IProductInCart = {
        title: item!.title,
        price: item!.price,
        image: item!.image,
        id: item!.id,
        category:item!.category,
        description: item!.description,
        counterIdentical: Number(quantity)
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart,
          newItem
        ],
        total: Number(((state.total) + (Number(newItem.price) * Number(quantity))).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) + Number(quantity)
      };

    case "ITEM_REMOVE_FROM_CART":
      const idx = action.payload as number;

      const itemIndex = state.itemsInCart.findIndex(item => item.id === idx);
      const itemDel = state.itemsInCart.find(item => item.id === idx);
      if(itemIndex === -1){
        return state;
      }
      return {
        ...state,
        itemsInCart: [
          ...state.itemsInCart.slice(0, itemIndex),
          ...state.itemsInCart.slice(itemIndex + 1)
        ],
        total: Number((state.total - (itemDel!.price * itemDel!.counterIdentical)).toFixed(2)),
        totalQuantity: Number(state.totalQuantity) - itemDel!.counterIdentical
      };

    case "CLEAR_CART":
      return {
        ...state,
        itemsInCart: [],
        total: 0,
        totalQuantity: 0
      }

    case "GET_AUTH_REQUEST":
      return {
        ...state,
        loading: true
      };

    case "GET_AUTH":
      return {
        ...state,
        auth: true,
        loading: false
      };

    case "GET_ADMIN_AUTH":
      return {
        ...state,
        auth: true,
        loading: false,
        admin: true
      };

    case "GET_AUTH_REJECT":
      return {
        ...state,
        loading: false
      };

    case "EXIT_AUTH":
      return {
        ...state,
        auth: false,
        admin: false
      };

    default:
      return state;
  };
};

export default reducer;