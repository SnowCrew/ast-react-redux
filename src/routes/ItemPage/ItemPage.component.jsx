import { connect } from "react-redux";
import {
  menuLoaded,
  menuRequested,
  menuError,
  onAddToCart,
} from "../../redux/actions/actions";
import { Link, useParams } from "react-router-dom";
import "./ItemPage.styles.css";
import { useEffect } from "react";
import WithAstService from "../../hoc/with-ast-service.hoc";
import Loading from "../../components/Loading/Loading.component";

const ItemPage = (props) => {
  const paramId = useParams().id;

  const {
    menuLoaded,
    menuError,
    AstService,
    menuRequested,
    products,
    loading,
  } = props;

  useEffect(() => {
    if (products.length === null) {
      menuRequested();

      AstService.getAllProducts()
        .then((res) => props.menuLoaded(res))
        .catch((err) => menuError());
    }
  }, [
    AstService,
    menuError,
    menuLoaded,
    menuRequested,
    products.length,
    props,
  ]);

  const Spinner = loading ? (
    <div className="item_page">
      <Loading />
    </div>
  ) : null;
  const item = props.products.find((el) => +el.id === +paramId);
  const { title, image, category, price, id } = item;

  return (
    <>
      {Spinner}
      <div className="item_page">
        <div className="menu__item item_block">
          <div className="menu__title">{title}</div>
          <img className="menu__img1" src={image} alt={title}></img>
          <div className="menu__category">
            Category: <span>{category}</span>
          </div>
          <div className="menu__price">
            Price: <span>{price}$</span>
          </div>
          <button onClick={() => onAddToCart(id)} className="menu__btn">
            Add to cart
          </button>
          <Link to={`/`}>
            <button className="menu__btn">Back</button>
          </Link>
          <span className={`menu__category_Img ${category}`}></span>
        </div>
      </div>
    </>
  );
};

const mapStateToProps = (state) => {
  return {
    products: state.products,
    loading: state.loading,
    error: state.error,
  };
};

const mapDispatchToProps = {
  menuLoaded,
  menuError,
  menuRequested,
  onAddToCart,
};

export default WithAstService()(
  connect(mapStateToProps, mapDispatchToProps)(ItemPage)
);
