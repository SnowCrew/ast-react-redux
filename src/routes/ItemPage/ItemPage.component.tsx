import SingleItemComponent from "../../components/SingleItem/SingleItem.component";

const ItemPage = () => {
  return <SingleItemComponent />;
};

export default ItemPage;
