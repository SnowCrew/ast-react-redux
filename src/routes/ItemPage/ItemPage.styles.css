.item_page {
  display: flex;

  max-width: 100vw;
  height: 100%;

  margin-bottom: 50px;
}
.item-block {
  display: flex;
  flex-direction: column;
  gap: 10px;

  margin: auto;
  width: 500px;
}

.item-page-img {
  width: 500px;
  height: 450px;
  transform: translateX(-20px);
  margin-bottom: 20px;
}

.menu__item {
  position: relative;
}
.menu__category_Img {
  position: absolute;
  content: "";
  bottom: 20px;
  right: 20px;
  width: 75px;
  height: 75px;
  background-position: 50%;
  background-size: cover;
  background-repeat: no-repeat;
}
.item-page-title {
  text-align: center;

  font-weight: bold;

  margin-bottom: 10px;
}

.item-page-quantitiy-container {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
}
.quantity-label {
  margin-right: 5px;
}

.textarea {
  resize: none;
}
