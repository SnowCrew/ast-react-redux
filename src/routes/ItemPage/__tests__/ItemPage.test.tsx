import { cleanup, render, screen } from "@testing-library/react";
import { ReactElement } from "react";
import { legacy_createStore as createStore } from "redux";
import reducer, { IProduct } from "../../../redux/reducers/reducer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom/extend-expect";
import ItemPage from "../ItemPage.component";

afterEach(cleanup);

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: "1",
  }),
}));

const renderWithRedux = (
  component: ReactElement,
  { initialState, store = createStore(reducer, initialState) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>, {
      wrapper: MemoryRouter,
    }),
  };
};
const randomProducts1: IProduct = {
  id: 1,
  title: "title1",
  price: 25,
  category: "category1",
  description: "description1",
  image: "image1",
};
const randomProducts2: IProduct = {
  id: 2,
  title: "title2",
  price: 25,
  category: "category2",
  description: "description12",
  image: "image2",
};

describe("ItemPage render without errors", () => {
  it("ItemPage rendering", () => {
    expect(
      renderWithRedux(<ItemPage />, {
        initialState: {
          products: [randomProducts1, randomProducts2],
          loading: false,
          error: false,
          itemsInCart: [],
          total: 0,
          totalQuantity: 0,
          auth: false,
          admin: false,
        },
      })
    ).toMatchSnapshot();
  });
});
