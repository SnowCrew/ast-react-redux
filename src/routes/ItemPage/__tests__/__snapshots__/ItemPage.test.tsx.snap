// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage render without errors CartPage rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="item_page"
      >
        <div
          class="item-block"
        >
          <div
            class="item-page-title"
          >
            title1
          </div>
          <img
            alt="title1"
            class="item-page-img"
            src="image1"
          />
          <div>
             
            description1
          </div>
          <div
            class="menu__category"
          >
            Category: 
            <span>
              category1
            </span>
          </div>
          <div
            class="menu__price"
          >
            Price: 
            <span>
              25
              $
            </span>
          </div>
          Чтобы добавить товар в корзину залогинтесь
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="item_page"
    >
      <div
        class="item-block"
      >
        <div
          class="item-page-title"
        >
          title1
        </div>
        <img
          alt="title1"
          class="item-page-img"
          src="image1"
        />
        <div>
           
          description1
        </div>
        <div
          class="menu__category"
        >
          Category: 
          <span>
            category1
          </span>
        </div>
        <div
          class="menu__price"
        >
          Price: 
          <span>
            25
            $
          </span>
        </div>
        Чтобы добавить товар в корзину залогинтесь
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ItemPage render without errors ItemPage rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="item_page"
      >
        <div
          class="item-block"
        >
          <div
            class="item-page-title"
          >
            title1
          </div>
          <img
            alt="title1"
            class="item-page-img"
            src="image1"
          />
          <div>
             
            description1
          </div>
          <div
            class="menu__category"
          >
            Category: 
            <span>
              category1
            </span>
          </div>
          <div
            class="menu__price"
          >
            Price: 
            <span>
              25
              $
            </span>
          </div>
          Чтобы добавить товар в корзину залогинтесь
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="item_page"
    >
      <div
        class="item-block"
      >
        <div
          class="item-page-title"
        >
          title1
        </div>
        <img
          alt="title1"
          class="item-page-img"
          src="image1"
        />
        <div>
           
          description1
        </div>
        <div
          class="menu__category"
        >
          Category: 
          <span>
            category1
          </span>
        </div>
        <div
          class="menu__price"
        >
          Price: 
          <span>
            25
            $
          </span>
        </div>
        Чтобы добавить товар в корзину залогинтесь
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
